test/unit/org/apache/cassandra/db/marshal/AbstractTypeTest.java

```diff
@@ -0,0 +1,439 @@
+import org.apache.cassandra.cql3.Constants;
+import org.apache.cassandra.cql3.Json;
+import org.apache.cassandra.cql3.QueryOptions;
+import org.apache.cassandra.cql3.Term;
+import org.apache.cassandra.cql3.VariableSpecifications;
+import org.apache.cassandra.io.util.DataInputBuffer;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputBuffer;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.schema.CQLTypeParser;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.Types;
+import org.apache.cassandra.transport.ProtocolVersion;
+import org.apache.cassandra.utils.AbstractTypeGenerators;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.FastByteOperations;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.apache.cassandra.utils.bytecomparable.ByteSourceInverse;
+import org.quicktheories.core.Gen;
+import org.reflections.Reflections;
+import org.reflections.scanners.Scanners;
+import org.reflections.util.ConfigurationBuilder;
+
+import static org.apache.cassandra.utils.AbstractTypeGenerators.TypeKind.*;
+import static org.apache.cassandra.utils.AbstractTypeGenerators.extractUDTs;
+import static org.apache.cassandra.utils.AbstractTypeGenerators.typeTree;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.quicktheories.QuickTheory.qt;
+
+public class AbstractTypeTest
+{
+    static
+    {
+        // make sure blob is always the same
+        CassandraRelevantProperties.TEST_BLOB_SHARED_SEED.setInt(42);
+    }
+
+    //TODO
+    // isCompatibleWith/isValueCompatibleWith/isSerializationCompatibleWith,
+    // withUpdatedUserType/expandUserTypes/referencesDuration - types that recursive check types
+    // getMaskedValue
+
+    @Test
+    public void allTypesCovered()
+    {
+        // this test just makes sure that all types are covered and no new type is left out
+        Reflections reflections = new Reflections(new ConfigurationBuilder()
+                                                  .forPackage("org.apache.cassandra")
+                                                  .setScanners(Scanners.SubTypes)
+                                                  .setExpandSuperTypes(true)
+                                                  .setParallel(true));
+        Set<Class<? extends AbstractType>> subTypes = reflections.getSubTypesOf(AbstractType.class);
+        Set<Class<? extends AbstractType>> coverage = AbstractTypeGenerators.knownTypes();
+        StringBuilder sb = new StringBuilder();
+        for (Class<? extends AbstractType> klass : Sets.difference(subTypes, coverage))
+        {
+            if (Modifier.isAbstract(klass.getModifiers()))
+                continue;
+            if ("test".equals(new File(klass.getProtectionDomain().getCodeSource().getLocation().getPath()).name()))
```

Comment:
if a test creates a type, doesn't matter if its *valid* as the test is likely testing its own internal logic.